hphp/util/asm-x64-intelxed.h

```diff
@@ -0,0 +1,1031 @@
+/*
+   +----------------------------------------------------------------------+
+   | HipHop for PHP                                                       |
+   +----------------------------------------------------------------------+
+   | Copyright (c) 2010-present Facebook, Inc. (http://www.facebook.com)  |
+   | Copyright (c) 2018 Intel Corporation                                 |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+*/
+#ifndef incl_HPHP_UTIL_ASM_X64_INTELXED_H_
+#define incl_HPHP_UTIL_ASM_X64_INTELXED_H_
+
+extern "C" {
+    #include <xed-interface.h>
+}
+/*
+ * A macro assembler for x64, based on the Intel XED library, that strives
+ * for low coupling to the runtime environment and ease of extendability.
+ */
+namespace HPHP { namespace jit {
+
+struct XedInit {
+  XedInit() {
+      xed_tables_init();
+  }
+};
+
+//////////////////////////////////////////////////////////////////////
+
+#define SZ_TO_BITS(sz)                    (sz << 3)
```

Comment:
I've changed `SZ_TO_BITS` and `BITS_TO_SZ` to `inline constexpr` functions in https://github.com/facebook/hhvm/pull/8164/commits/b0f08a27f56e94240848f527b4f5c87c043aa306. For `XED_REG(...)` and similar macros (lines 169 - 174), there are several approaches I could go for:
- leave them unchanged - I think it makes the code more readable
- change them to inline functions - adds more code that doesn't help with neither readability nor functionality
- replace the macros with their content - since each of them is a one-liner, it would not make the code too bloated, but I think it will still be less readable
- drop the `XedOperand` class, since all it does is conversion from HHVM objects (such as `Reg8/16/32/64`, `Immed`, `Immed64`) to a `xed_encoder_operand_t` instance which it contains as a class member, and replace it with a set of conversion functions that take one of the mentioned HHVM objects and return a `xed_encoder_operand_t`

Which one would you prefer?

Reply:
How would making them inline functions add much more code? The macros do have shorter definitions, but inline functions wouldn't require any more code at each usage site, and they're generally much better when it comes to compiler errors and debugging. These macros aren't doing anything weird like concatenating tokens together or referencing a variable not passed as an argument. `XED_REG` could be something like:
```c++
template<typename Reg>
xed_encoder_operand_t xed_reg(const Reg& r) {
  return XedOperand(r).op;
}

// Then, unless I'm missing something, you can call xed_reg just like you'd
// use the XED_REG macro:
XED_REG(blah)
// becomes
xed_reg(blah)
```
But then at that point you're not too far away from your last suggestion of replacing `XedOperand` with a set of conversion functions that don't use an intermediate object. I think going all the way there would be best.